File: metar_decode.js
function decodeMetar(text) {
    if (typeof text !== "string") return null;
    const rawTokens = text.trim().split(/\s+/);
    if (rawTokens.length === 0 || (rawTokens.length === 1 && rawTokens[0] === "")) return null;
    let tokens = rawTokens.slice();
    if (tokens[0] === "METAR" || tokens[0] === "SPECI") tokens = tokens.slice(1);
    if (tokens.length < 2) return null;
    const station = tokens[0];
    if (!/^[A-Z0-9]{4}$/.test(station)) return null;
    const timeMatch = tokens[1].match(/^(\d{2})(\d{2})(\d{2})Z$/);
    if (!timeMatch) return null;
    const time_z = timeMatch[2] + ":" + timeMatch[3] + "Z";
    const stopTokens = new Set(["RMK", "TEMPO", "BECMG", "PROB30", "PROB40"]);
    const body = [];
    for (let i = 2; i < tokens.length; i++) {
        const t = tokens[i];
        if (stopTokens.has(t) || /^FM\d+/.test(t)) break;
        if (t === "AUTO" || t === "COR" || t === "NIL") continue;
        body.push(t);
    }
    const result = {
        station: station,
        time_z: time_z,
        wind_dir: null,
        wind_variable: false,
        wind_speed_kt: null,
        wind_gust_kt: null,
        wind_var_from: null,
        wind_var_to: null,
        cavok: false,
        visibility_m: null,
        clouds: [],
        ceiling_ft: null,
        weather: [],
        temp_c: null,
        dew_c: null,
        altimeter_inhg: null,
        qnh_hpa: null,
        category: null
    };
    // WIND
    for (const t of body) {
        const m = t.match(/^(VRB|\d{3})(\d{2,3})(?:G(\d{2,3}))?(KT|MPS|KMH)$/);
        if (m) {
            if (m[1] === "VRB") {
                result.wind_variable = true;
                result.wind_dir = null;
            } else {
                result.wind_dir = parseInt(m[1], 10);
            }
            const speedVal = parseInt(m[2], 10);
            const gustVal = m[3] ? parseInt(m[3], 10) : null;
            const unit = m[4];
            const toKnots = (val) => {
                if (unit === "KT") return val;
                if (unit === "MPS") return Math.floor(val * 1.943844 + 0.5);
                return Math.floor(val / 1.852 + 0.5);
            };
            result.wind_speed_kt = toKnots(speedVal);
            result.wind_gust_kt = gustVal !== null ? toKnots(gustVal) : null;
            break;
        }
    }
    // WIND VARIATION
    for (const t of body) {
        const m = t.match(/^(\d{3})V(\d{3})$/);
        if (m) {
            result.wind_var_from = parseInt(m[1], 10);
            result.wind_var_to = parseInt(m[2], 10);
            break;
        }
    }
    // CAVOK
    if (body.includes("CAVOK")) {
        result.cavok = true;
        result.visibility_m = 10000;
    }
    // VISIBILITY
    if (!result.cavok) {
        for (let i = 0; i < body.length; i++) {
            const t = body[i];
            let m;
            if (/^\d{4}$/.test(t)) {
                result.visibility_m = parseInt(t, 10);
                break;
            }
            if ((m = t.match(/^(\d+)SM$/))) {
                const sm = parseInt(m[1], 10);
                result.visibility_m = Math.floor(sm * 1609.344 + 0.5);
                break;
            }
            if ((m = t.match(/^P(\d+)SM$/))) {
                const sm = parseInt(m[1], 10);
                result.visibility_m = Math.floor(sm * 1609.344 + 0.5);
                break;
            }
            if ((m = t.match(/^M(\d)\/(\d)SM$/))) {
                const sm = parseInt(m[1], 10) / parseInt(m[2], 10);
                result.visibility_m = Math.floor(sm * 1609.344 + 0.5);
                break;
            }
            if ((m = t.match(/^(\d)\/(\d)SM$/))) {
                const sm = parseInt(m[1], 10) / parseInt(m[2], 10);
                result.visibility_m = Math.floor(sm * 1609.344 + 0.5);
                break;
            }
            if (/^\d$/.test(t) && i + 1 < body.length && (m = body[i + 1].match(/^(\d)\/(\d)SM$/))) {
                const whole = parseInt(t, 10);
                const frac = parseInt(m[1], 10) / parseInt(m[2], 10);
                const sm = whole + frac;
                result.visibility_m = Math.floor(sm * 1609.344 + 0.5);
                break;
            }
        }
    }
    // CLOUDS and CEILING
    let vvFt = null;
    for (const t of body) {
        let m;
        if ((m = t.match(/^(FEW|SCT|BKN|OVC)(\d{3})(CB|TCU)?$/))) {
            const cover = m[1];
            const height_ft = parseInt(m[2], 10) * 100;
            const type = m[3] || null;
            result.clouds.push({ cover, height_ft, type });
            if (cover === "BKN" || cover === "OVC") {
                if (result.ceiling_ft === null || height_ft < result.ceiling_ft) {
                    result.ceiling_ft = height_ft;
                }
            }
        } else if ((m = t.match(/^VV(\d{3})$/))) {
            vvFt = parseInt(m[1], 10) * 100;
        }
    }
    if (vvFt !== null) {
        if (result.ceiling_ft === null || vvFt < result.ceiling_ft) {
            result.ceiling_ft = vvFt;
        }
    }
    // TEMPERATURE
    for (const t of body) {
        const m = t.match(/^(M?\d{2})\/(M?\d{2})?$/);
        if (m) {
            const parseTemp = (s) => {
                if (s.startsWith("M")) return -parseInt(s.slice(1), 10);
                return parseInt(s, 10);
            };
            result.temp_c = parseTemp(m[1]);
            result.dew_c = m[2] ? parseTemp(m[2]) : null;
            break;
        }
    }
    // PRESSURE
    for (const t of body) {
        let m;
        if ((m = t.match(/^Q(\d{4})$/))) {
            const qnh = parseInt(m[1], 10);
            result.qnh_hpa = qnh;
            result.altimeter_inhg = Math.round(qnh / 33.8639 * 100) / 100;
            break;
        }
        if ((m = t.match(/^A(\d{4})$/))) {
            const alt = parseInt(m[1], 10) / 100;
            result.altimeter_inhg = alt;
            result.qnh_hpa = Math.floor(alt * 33.8639 + 0.5);
            break;
        }
    }
    // WEATHER
    for (const t of body) {
        if (/^[-+]?(VC)?(MI|BC|PR|DR|BL|SH|TS|FZ)?(DZ|RA|SN|SG|IC|PL|GR|GS|UP|BR|FG|FU|VA|DU|SA|HZ|PY|PO|SQ|FC|SS|DS)+$/.test(t)) {
            result.weather.push(t);
        }
    }
    // CATEGORY
    const visSm = result.visibility_m !== null ? Math.round(result.visibility_m / 1609.344 * 100) / 100 : null;
    const visCat = visSm === null ? "VFR" : visSm < 1 ? "LIFR" : visSm < 3 ? "IFR" : visSm <= 5 ? "MVFR" : "VFR";
    const ceilCat = result.ceiling_ft === null ? "VFR" : result.ceiling_ft < 500 ? "LIFR" : result.ceiling_ft < 1000 ? "IFR" : result.ceiling_ft <= 3000 ? "MVFR" : "VFR";
    if (result.cavok) {
        result.category = "VFR";
    } else if (result.visibility_m === null && result.ceiling_ft === null && result.clouds.length === 0) {
        result.category = null;
    } else {
        const order = { VFR: 0, MVFR: 1, IFR: 2, LIFR: 3 };
        result.category = order[visCat] > order[ceilCat] ? visCat : ceilCat;
    }
    return result;
}

File: metar_decode.py
import re

def decode_metar(text):
    if not isinstance(text, str):
        return None
    tokens = text.split()
    if tokens and tokens[0] in ("METAR", "SPECI"):
        tokens = tokens[1:]
    if len(tokens) < 2:
        return None
    station = tokens[0]
    if not re.fullmatch(r'^[A-Z0-9]{4}$', station):
        return None
    if not re.fullmatch(r'^(\d{2})(\d{2})(\d{2})Z$', tokens[1]):
        return None
    m = re.fullmatch(r'^(\d{2})(\d{2})(\d{2})Z$', tokens[1])
    dd, hh, mm = m.group(1), m.group(2), m.group(3)
    time_z = hh + ":" + mm + "Z"
    body = tokens[2:]
    cut_idx = len(body)
    for i, tok in enumerate(body):
        if tok in ("RMK", "TEMPO", "BECMG", "PROB30", "PROB40") or re.fullmatch(r'^FM\d+', tok):
            cut_idx = i
            break
    body = body[:cut_idx]
    result = {
        "station": station,
        "time_z": time_z,
        "wind_dir": None,
        "wind_variable": False,
        "wind_speed_kt": None,
        "wind_gust_kt": None,
        "wind_var_from": None,
        "wind_var_to": None,
        "cavok": False,
        "visibility_m": None,
        "clouds": [],
        "ceiling_ft": None,
        "weather": [],
        "temp_c": None,
        "dew_c": None,
        "altimeter_inhg": None,
        "qnh_hpa": None,
        "category": None
    }
    wind_re = re.compile(r'^(VRB|\d{3})(\d{2,3})(?:G(\d{2,3}))?(KT|MPS|KMH)$')
    for tok in body:
        m = wind_re.fullmatch(tok)
        if m:
            g1, g2, g3, unit = m.group(1), m.group(2), m.group(3), m.group(4)
            if g1 == "VRB":
                result["wind_variable"] = True
                result["wind_dir"] = None
            else:
                result["wind_dir"] = int(g1)
            speed = int(g2)
            gust = int(g3) if g3 is not None else None
            if unit == "KT":
                pass
            elif unit == "MPS":
                speed = int(speed * 1.943844 + 0.5)
                if gust is not None:
                    gust = int(gust * 1.943844 + 0.5)
            elif unit == "KMH":
                speed = int(speed / 1.852 + 0.5)
                if gust is not None:
                    gust = int(gust / 1.852 + 0.5)
            result["wind_speed_kt"] = speed
            result["wind_gust_kt"] = gust
            break
    var_re = re.compile(r'^(\d{3})V(\d{3})$')
    for tok in body:
        m = var_re.fullmatch(tok)
        if m:
            result["wind_var_from"] = int(m.group(1))
            result["wind_var_to"] = int(m.group(2))
            break
    if "CAVOK" in body:
        result["cavok"] = True
        result["visibility_m"] = 10000
    else:
        vis_found = False
        i = 0
        while i < len(body) and not vis_found:
            tok = body[i]
            if re.fullmatch(r'^\d{4}$', tok):
                result["visibility_m"] = int(tok)
                vis_found = True
            elif re.fullmatch(r'^(\d+)SM$', tok):
                sm = int(re.fullmatch(r'^(\d+)SM$', tok).group(1))
                result["visibility_m"] = int(sm * 1609.344 + 0.5)
                vis_found = True
            elif re.fullmatch(r'^P(\d+)SM$', tok):
                sm = int(re.fullmatch(r'^P(\d+)SM$', tok).group(1))
                result["visibility_m"] = int(sm * 1609.344 + 0.5)
                vis_found = True
            elif re.fullmatch(r'^M(\d)/(\d)SM$', tok):
                m = re.fullmatch(r'^M(\d)/(\d)SM$', tok)
                sm = int(m.group(1)) / int(m.group(2))
                result["visibility_m"] = int(sm * 1609.344 + 0.5)
                vis_found = True
            elif re.fullmatch(r'^(\d)/(\d)SM$', tok):
                m = re.fullmatch(r'^(\d)/(\d)SM$', tok)
                sm = int(m.group(1)) / int(m.group(2))
                result["visibility_m"] = int(sm * 1609.344 + 0.5)
                vis_found = True
            elif re.fullmatch(r'^\d$', tok) and i + 1 < len(body) and re.fullmatch(r'^(\d)/(\d)SM$', body[i + 1]):
                whole = int(tok)
                m = re.fullmatch(r'^(\d)/(\d)SM$', body[i + 1])
                sm = whole + int(m.group(1)) / int(m.group(2))
                result["visibility_m"] = int(sm * 1609.344 + 0.5)
                vis_found = True
                i += 1
            i += 1
    cloud_re = re.compile(r'^(FEW|SCT|BKN|OVC)(\d{3})(CB|TCU)?$')
    vv_re = re.compile(r'^VV(\d{3})$')
    vv_ft = None
    for tok in body:
        m = cloud_re.fullmatch(tok)
        if m:
            result["clouds"].append({
                "cover": m.group(1),
                "height_ft": int(m.group(2)) * 100,
                "type": m.group(3)
            })
        m = vv_re.fullmatch(tok)
        if m:
            vv_ft = int(m.group(1)) * 100
    heights = [c["height_ft"] for c in result["clouds"] if c["cover"] in ("BKN", "OVC")]
    if vv_ft is not None:
        heights.append(vv_ft)
    if heights:
        result["ceiling_ft"] = min(heights)
    temp_re = re.compile(r'^(M?\d{2})/(M?\d{2})?$')
    for tok in body:
        m = temp_re.fullmatch(tok)
        if m:
            t = m.group(1)
            d = m.group(2)
            if t.startswith("M"):
                result["temp_c"] = -int(t[1:])
            else:
                result["temp_c"] = int(t)
            if d is not None:
                if d.startswith("M"):
                    result["dew_c"] = -int(d[1:])
                else:
                    result["dew_c"] = int(d)
            break
    q_re = re.compile(r'^Q(\d{4})$')
    a_re = re.compile(r'^A(\d{4})$')
    for tok in body:
        m = q_re.fullmatch(tok)
        if m:
            result["qnh_hpa"] = int(m.group(1))
            result["altimeter_inhg"] = round(result["qnh_hpa"] / 33.8639, 2)
            break
        m = a_re.fullmatch(tok)
        if m:
            result["altimeter_inhg"] = int(m.group(1)) / 100
            result["qnh_hpa"] = int(result["altimeter_inhg"] * 33.8639 + 0.5)
            break
    wx_re = re.compile(r'^[-+]?(VC)?(MI|BC|PR|DR|BL|SH|TS|FZ)?(DZ|RA|SN|SG|IC|PL|GR|GS|UP|BR|FG|FU|VA|DU|SA|HZ|PY|PO|SQ|FC|SS|DS)+$')
    for tok in body:
        if wx_re.fullmatch(tok):
            result["weather"].append(tok)
    # rounded so that exactly 3 SM (4828 m) counts as 3.0 and not as 2.9999
    vis_sm = round(result["visibility_m"] / 1609.344, 2) if result["visibility_m"] is not None else None
    if vis_sm is not None:
        if vis_sm < 1:
            vis_cat = "LIFR"
        elif vis_sm < 3:
            vis_cat = "IFR"
        elif vis_sm <= 5:
            vis_cat = "MVFR"
        else:
            vis_cat = "VFR"
    else:
        vis_cat = "VFR"
    if result["ceiling_ft"] is not None:
        if result["ceiling_ft"] < 500:
            ceil_cat = "LIFR"
        elif result["ceiling_ft"] < 1000:
            ceil_cat = "IFR"
        elif result["ceiling_ft"] <= 3000:
            ceil_cat = "MVFR"
        else:
            ceil_cat = "VFR"
    else:
        ceil_cat = "VFR"
    if result["visibility_m"] is None and result["ceiling_ft"] is None and len(result["clouds"]) == 0:
        if result["cavok"]:
            result["category"] = "VFR"
        else:
            result["category"] = None
    else:
        order = ["VFR", "MVFR", "IFR", "LIFR"]
        idx_vis = order.index(vis_cat)
        idx_ceil = order.index(ceil_cat)
        result["category"] = order[max(idx_vis, idx_ceil)]
    return result